tests/fuzzers/bls12381/bls12381_fuzz.go

```diff
@@ -0,0 +1,250 @@
+
+	// get random G1 points
+	kpG1, cpG1, err := getG1Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// get random G2 points
+	kpG2, cpG2, err := getG2Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// compute pairing using geth
+	engine := bls12381.NewPairingEngine()
+	engine.AddPair(kpG1, kpG2)
+	kResult := engine.Result()
+
+	// compute pairing using gnark
+	cResult, err := gnark.Pair([]gnark.G1Affine{*cpG1}, []gnark.G2Affine{*cpG2})
+	if err != nil {
+		panic(fmt.Sprintf("gnark/bls12381 encountered error: %v", err))
+	}
+
+	// compare result
+	if !(bytes.Equal(cResult.Marshal(), bls12381.NewGT().ToBytes(kResult))) {
+		panic("pairing mismatch gnark / geth ")
+	}
+
+	return 1
+}
+
+func FuzzCrossG1Add(data []byte) int {
+	input := bytes.NewReader(data)
+
+	// get random G1 points
+	kp1, cp1, err := getG1Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// get random G1 points
+	kp2, cp2, err := getG1Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// compute kp = kp1 + kp2
+	g1 := bls12381.NewG1()
+	kp := bls12381.PointG1{}
+	g1.Add(&kp, kp1, kp2)
+
+	// compute cp = cp1 + cp2
+	_cp1 := new(gnark.G1Jac).FromAffine(cp1)
+	_cp2 := new(gnark.G1Jac).FromAffine(cp2)
+	cp := new(gnark.G1Affine).FromJacobian(_cp1.AddAssign(_cp2))
+
+	// compare result
+	if !(bytes.Equal(cp.Marshal(), g1.ToBytes(&kp))) {
+		panic("G1 point addition mismatch gnark / geth ")
+	}
+
+	return 1
+}
+
+func FuzzCrossG2Add(data []byte) int {
+	input := bytes.NewReader(data)
+
+	// get random G2 points
+	kp1, cp1, err := getG2Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// get random G2 points
+	kp2, cp2, err := getG2Points(input)
+	if err != nil {
+		return 0
+	}
+
+	// compute kp = kp1 + kp2
+	g2 := bls12381.NewG2()
+	kp := bls12381.PointG2{}
+	g2.Add(&kp, kp1, kp2)
+
+	// compute cp = cp1 + cp2
+	_cp1 := new(gnark.G2Jac).FromAffine(cp1)
+	_cp2 := new(gnark.G2Jac).FromAffine(cp2)
+	cp := new(gnark.G2Affine).FromJacobian(_cp1.AddAssign(_cp2))
+
+	// compare result
+	if !(bytes.Equal(cp.Marshal(), g2.ToBytes(&kp))) {
+		panic("G2 point addition mismatch gnark / geth ")
+	}
+
+	return 1
+}
+
+func FuzzCrossG1MultiExp(data []byte) int {
+	// result = multiExp(bases, scalars)
+	// n random scalars
+
+	// we take the first byte to derive n (max = 17)
+	if len(data) < 1 {
+		return 0
+	}
+	n := (uint8(data[0]) % 16) + 1
+
+	input := bytes.NewReader(data[1:])
+
+	// n random scalars
+	scalars := make([]*big.Int, n)
+	_scalars := make([]fr.Element, n)
+	for i := 0; i < int(n); i++ {
+		// note that geth/crypto/bls12381 works only with scalars <= 32bytes
+		s, err := randomScalar(input, fr.Modulus())
+		if err != nil {
+			return 0
+		}
+		scalars[i] = s
+		_scalars[i].SetBigInt(scalars[i]).FromMont()
+	}
+
+	// get a random G1 point as basis
+	kp1, cp1, err := getG1Points(input)
+	if err != nil {
+		return 0
+	}
+	kps := make([]*bls12381.PointG1, n)
+	cps := make([]gnark.G1Affine, n)
+	for i := 0; i < int(n); i++ {
+		kps[i] = new(bls12381.PointG1).Set(kp1)
```

Comment:
I would prefer setting the G1 points to different points, not only to a single point.

Reply:
I've fixed that now -- it reads as many as it can (max 17), and uses those, and does not use the `n`